curated_transformers/util/auto_model.py

```diff
@@ -0,0 +1,227 @@
+
+    _HF_MODEL_TYPE_TO_CURATED = {
+        "bert": BertEncoder,
+        "albert": AlbertEncoder,
+        "camembert": CamembertEncoder,
+        "roberta": RobertaEncoder,
+        "xlm-roberta": XlmRobertaEncoder,
+    }
+
+    @classmethod
+    def from_hf_hub(
+        cls,
+        name: str,
+        revision: str = "main",
+        *,
+        device: Optional[torch.device] = None,
+        quantization_config: Optional[BitsAndBytesConfig] = None,
+    ) -> EncoderModule:
+        encoder = _instantiate_module_from_hf_hub(
+            name,
+            revision,
+            device,
+            quantization_config,
+            cls._HF_MODEL_TYPE_TO_CURATED,
+            "encoder",
+        )
+        assert isinstance(encoder, EncoderModule)
+        return encoder
+
+
+class AutoDecoder(AutoModel[DecoderModule]):
+    """Decoder module loaded from the Hugging Face Model Hub."""
+
+    _HF_MODEL_TYPE_TO_CURATED = {
+        "gpt_neox": GPTNeoXDecoder,
+        "RefinedWeb": RefinedWebModelDecoder,
+        "RefinedWebModel": RefinedWebModelDecoder,
+    }
+
+    @classmethod
+    def from_hf_hub(
+        cls,
+        name: str,
+        revision: str = "main",
+        *,
+        device: Optional[torch.device] = None,
+        quantization_config: Optional[BitsAndBytesConfig] = None,
+    ) -> DecoderModule:
+        decoder = _instantiate_module_from_hf_hub(
+            name,
+            revision,
+            device,
+            quantization_config,
+            cls._HF_MODEL_TYPE_TO_CURATED,
+            "decoder",
+        )
+        assert isinstance(decoder, DecoderModule)
+        return decoder
+
+
+class AutoCausalLM(AutoModel[CausalLMModule[KeyValueCache]]):
+    """Causal LM module loaded from the Hugging Face Model Hub."""
+
+    _HF_MODEL_TYPE_TO_CURATED = {
+        "gpt_neox": GPTNeoXCausalLM,
+        "RefinedWeb": RefinedWebModelCausalLM,
+        "RefinedWebModel": RefinedWebModelCausalLM,
+    }
+
+    @classmethod
+    def from_hf_hub(
+        cls,
+        name: str,
+        revision: str = "main",
+        *,
+        device: Optional[torch.device] = None,
+        quantization_config: Optional[BitsAndBytesConfig] = None,
+    ) -> CausalLMModule[KeyValueCache]:
+        causal_lm = _instantiate_module_from_hf_hub(
+            name,
+            revision,
+            device,
+            quantization_config,
+            cls._HF_MODEL_TYPE_TO_CURATED,
+            "causal LM",
+        )
+        assert isinstance(causal_lm, CausalLMModule)
+        return causal_lm
+
+
+class AutoGenerator(AutoModel[GeneratorWrapper]):
+    """Causal LM generator loaded from the Hugging Face Model Hub.
+
+    **NOTE** - This class can currently only be used with following models:
+        - ``databricks/dolly-v2`` variants.
+        - ``tiiuae/falcon`` variants.
+    """
+
+    _DOLLY_V2_PREFIX = "databricks/dolly-v2"
+    _FALCON_PREFIX = "tiiuae/falcon"
```

Comment:
I think we should match for `dolly-v2` and `falcon` occurring in the string. Usually there are a lot of custom finetuned flavors and it would be a shame if they can't be used.